// @sap-ux/fe-mockserver-core/dist/request/odataRequest.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const searchParser_1 = require("./searchParser");
const filterParser_1 = require("./filterParser");
const common_1 = require("../data/common");
const balanced_match_1 = __importDefault(require("balanced-match"));
const query_string_1 = require("query-string");
const url_1 = require("url");
class ODataRequest {
    constructor(requestContent, dataAccess) {
        var _a;
        this.requestContent = requestContent;
        this.dataAccess = dataAccess;
        this.responseHeaders = {};
        this.statusCode = 200;
        this.messages = [];
        const decodedUrl = decodeURIComponent(requestContent.url);
        const parsedUrl = new url_1.URL(`http://dummy${decodedUrl}`);
        const undecodedParsedUrl = new url_1.URL(`http://dummy${requestContent.url}`);
        this.tenantId = requestContent.tenantId || 'tenant-default';
        this.context = requestContent.url.split('?')[0].substring(1);
        if (this.tenantId) {
            this.addResponseHeader('sap-tenantid', this.tenantId);
        }
        this.isMinimalRepresentation = ((_a = requestContent.headers) === null || _a === void 0 ? void 0 : _a['prefer']) === 'return=minimal';
        this.queryPath = this.parsePath(undecodedParsedUrl.pathname.substring(1));
        this.parseParameters(parsedUrl.searchParams);
    }
    parseParameters(searchParams) {
        this.allParams = searchParams;
        this.searchQuery = (0, searchParser_1.parseSearch)(searchParams.get('$search'));
        this.orderBy = this.parseOrderBy(searchParams.get('$orderby'));
        this.startIndex = parseInt(searchParams.get('$skip') || '', 10);
        if (isNaN(this.startIndex)) {
            this.startIndex = 0;
        }
        this.maxElements = parseInt(searchParams.get('$top') || Number.POSITIVE_INFINITY.toString(), 10);
        if (isNaN(this.maxElements)) {
            this.maxElements = Number.POSITIVE_INFINITY;
        }
        this.aggregateDefinition = this.parseApply(searchParams.get('$apply'));
        this.filterDefinition = (0, filterParser_1.parseFilter)(searchParams.get('$filter'));
        this.countRequested = searchParams.has('$count');
        this.selectedProperties = {};
        const selectParams = searchParams.get('$select');
        const expandParams = searchParams.get('$expand');
        if (selectParams) {
            const props = selectParams.split(',');
            props.forEach((property) => {
                if (property.length > 0) {
                    this.selectedProperties[property.split('/')[0]] = true;
                }
            });
        }
        else {
            this.selectedProperties['*'] = true;
        }
        if (expandParams) {
            const expandParameters = this.parseExpand(expandParams);
            this.expandProperties = expandParameters.expand;
            this.selectedProperties = Object.assign(this.selectedProperties, expandParameters.properties || {});
        }
        if (this.aggregateDefinition) {
            const additionalSelectProperty = {};
            this.aggregateDefinition.groupBy.forEach((groupByProperty) => {
                additionalSelectProperty[groupByProperty] = true;
            });
            this.aggregateDefinition.aggregates.forEach((aggregateSourceProp) => {
                additionalSelectProperty[aggregateSourceProp.sourceProperty] = true;
            });
            this.selectedProperties = Object.assign(this.selectedProperties, additionalSelectProperty);
        }
    }
    /**
     * Split a list of properties that can contain sub-requests into an array.
     *
     * @param propertiesQuery OData properties request
     * @param delimiter Property delimiter
     * @returns an array of string with the properties
     */
    splitProperties(propertiesQuery, delimiter = ',') {
        const properties = [];
        let nestingDepth = 0, startIndex = 0, index = 0;
        for (; index < propertiesQuery.length; index++) {
            const char = propertiesQuery[index];
            if (char === delimiter && nestingDepth === 0) {
                // top-level delimiter — end of property
                if (index - startIndex > 0) {
                    properties.push(propertiesQuery.substring(startIndex, index));
                }
                startIndex = index + 1;
            }
            else if (char === '(') {
                nestingDepth++;
            }
            else if (char === ')') {
                nestingDepth--;
            }
        }
        if (index - startIndex > 0) {
            properties.push(propertiesQuery.substring(startIndex));
        }
        if (nestingDepth !== 0) {
            throw new common_1.ExecutionError(`Too many ${nestingDepth > 0 ? 'opening' : 'closing'} parentheses: ${propertiesQuery}`, 400, undefined, false);
        }
        return properties;
    }
    parseExpand(expandParameters) {
        const props = this.splitProperties(expandParameters);
        return props.reduce((reducer, property) => {
            var _a;
            if (this.dataAccess.getMetadata().getVersion() === '4.0') {
                const { pre: name, body: parameters } = (_a = (0, balanced_match_1.default)('(', ')', property)) !== null && _a !== void 0 ? _a : { pre: property, body: '' };
                const parameterSplit = this.splitProperties(parameters, ';');
                const queryPart = parameterSplit.reduce((acc, split) => Object.assign(acc, (0, query_string_1.parse)(split)), {});
                const expand = queryPart['$expand']
                    ? this.parseExpand(queryPart['$expand'])
                    : { expand: {}, properties: {} };
                const selectProperties = {};
                if (queryPart['$select']) {
                    // explicit $select
                    queryPart['$select']
                        .split(',')
                        .forEach((propertyName) => (selectProperties[propertyName] = true));
                }
                else {
                    selectProperties['*'] = true;
                }
                Object.keys(expand.properties).forEach((expandName) => {
                    selectProperties[expandName] = true;
                });
                reducer.expand[name] = {
                    expand: expand.expand,
                    properties: selectProperties
                };
                reducer.properties[name] = true;
                return reducer;
            }
            else {
                const propertySplit = property.split('/');
                const name = propertySplit[0];
                const expand = propertySplit[1]
                    ? this.parseExpand(propertySplit[1])
                    : { expand: {}, properties: {} };
                reducer.expand[name] = {
                    expand: expand.expand,
                    properties: {
                        '*': true
                    }
                };
                reducer.properties[name] = true;
                return reducer;
            }
        }, { expand: {}, properties: {} });
    }
    parseOrderBy(orderByParameters) {
        if (!orderByParameters) {
            return [];
        }
        const orderByParams = orderByParameters.split(',');
        const orderByDefinition = [];
        orderByParams.forEach((param) => {
            const [paramName, direction] = param.split(' ');
            const realDirection = ['asc', 'desc'].includes(direction)
                ? direction
                : 'asc';
            orderByDefinition.push({ name: paramName, direction: realDirection });
        });
        return orderByDefinition;
    }
    parsePath(path) {
        const pathSplit = path.split('/');
        return pathSplit.reduce((pathArr, pathPart) => {
            const keysStart = pathPart.indexOf('(');
            const keysEnd = pathPart.indexOf(')');
            let entity;
            let keys = {};
            if (keysStart > -1) {
                entity = pathPart.substring(0, keysStart) + pathPart.substring(keysEnd + 1);
                const keysList = pathPart.substring(keysStart + 1, keysEnd).split(',');
                keys = {};
                keysList.forEach((keyValue) => {
                    const [key, value] = keyValue.split('=');
                    if (value) {
                        keys[key] = decodeURIComponent(value.replace(/^(?:')$/g, ''));
                    }
                    else {
                        keys[key] = undefined;
                    }
                });
            }
            else {
                entity = pathPart;
            }
            pathArr.push({ path: entity, keys: keys });
            return pathArr;
        }, []);
    }
    parseApply(applyParameters) {
        if (!applyParameters) {
            return undefined;
        }
        const filterRegEx = /^filter\(([^)]+)\)\/(.*)$/;
        const filterMatches = applyParameters.match(filterRegEx);
        let groupByText = applyParameters;
        let filterParams;
        if (filterMatches) {
            const filterExpr = filterMatches[1];
            filterParams = (0, filterParser_1.parseFilter)(filterExpr);
            groupByText = filterMatches[2];
        }
        const groupByRegEx = /^groupby\(\(([^)]+)\),([^)]+\))\)$/;
        const groupByMatches = groupByText.match(groupByRegEx);
        if (groupByMatches) {
            return {
                filter: filterParams,
                groupBy: groupByMatches[1].split(','),
                aggregates: this.parseAggregateDefinition(groupByMatches[2])
            };
        }
    }
    parseAggregateDefinition(aggregationDefinition) {
        const aggregateRegEx = /^aggregate\(([^)]+)\)$/;
        const aggregateMatches = aggregationDefinition.match(aggregateRegEx);
        if (aggregateMatches) {
            return aggregateMatches[1].split(',').map((aggregateMatch) => {
                const aggregateSplit = aggregateMatch.split(' ');
                const property = aggregateSplit[0];
                const operator = aggregateSplit[2];
                const targetName = aggregateSplit[4];
                return {
                    name: targetName || property,
                    operator,
                    sourceProperty: property
                };
            });
        }
        else {
            return [];
        }
    }
    async handleRequest() {
        var _a;
        const contextId = (_a = this.requestContent.headers) === null || _a === void 0 ? void 0 : _a['sap-contextid'];
        if (contextId) {
            this.dataAccess.resetStickySessionTimeout(this, this.tenantId);
        }
        try {
            switch (this.requestContent.method) {
                case 'PATCH':
                case 'MERGE':
                case 'PUT': {
                    const updatedData = await this.dataAccess.updateData(this, this.requestContent.body);
                    this.setResponseData(updatedData);
                    break;
                }
                case 'DELETE':
                    await this.dataAccess.deleteData(this);
                    this.statusCode = 204;
                    break;
                case 'POST': {
                    const actionResponse = await this.dataAccess.performAction(this, this.requestContent.body);
                    if (actionResponse === null) {
                        this.setResponseData(await this.dataAccess.createData(this, this.requestContent.body));
                        this.statusCode = 201;
                    }
                    else {
                        this.setResponseData(actionResponse);
                    }
                    break;
                }
                case 'HEAD':
                    break; // do nothing
                case 'GET':
                default: {
                    const actionResponse = await this.dataAccess.performAction(this);
                    if (actionResponse === null) {
                        const retrievedData = await this.dataAccess.getData(this);
                        if (retrievedData === undefined || retrievedData === null) {
                            this.statusCode = 404;
                            this.setResponseData('');
                        }
                        else {
                            this.setResponseData(retrievedData);
                        }
                    }
                    else {
                        this.setResponseData(actionResponse);
                    }
                    break;
                }
            }
        }
        catch (errorInfo) {
            const errorInformation = errorInfo;
            if (errorInformation.isCustomError) {
                if (errorInformation.messageData) {
                    if (errorInformation.isSAPMessage) {
                        this.addResponseHeader('sap-messages', JSON.stringify(errorInformation.messageData));
                    }
                    else {
                        this.addResponseHeader('content-type', 'application/json;odata.metadata=minimal;IEEE754Compatible=true');
                        this.setResponseData(JSON.stringify(errorInformation.messageData));
                    }
                }
                else {
                    this.addResponseHeader('content-type', 'text/plain');
                    this.setResponseData(errorInformation.message);
                }
                if (Object.keys(errorInformation.headers).length > 0) {
                    Object.keys(errorInformation.headers).forEach((headerName) => {
                        this.addResponseHeader(headerName, errorInformation.headers[headerName]);
                    });
                }
                this.statusCode = errorInformation.statusCode;
            }
            else {
                this.statusCode = 500;
                this.addResponseHeader('content-type', 'text/plain');
                this.setResponseData(errorInformation.message);
            }
        }
    }
    setResponseData(data) {
        this.responseData = data;
    }
    setContext(context) {
        this.context = context;
    }
    getResponseData() {
        if (this.messages.length) {
            this.addResponseHeader('sap-messages', JSON.stringify(this.messages));
        }
        if (this.dataAccess.getMetadata().getVersion() === '4.0') {
            this.addResponseHeader('odata-version', '4.0');
        }
        else {
            this.addResponseHeader('dataserviceversion', '2.0');
            this.addResponseHeader('cache-control', 'no-store, no-cache');
            this.addResponseHeader('content-type', 'application/json');
        }
        if (typeof this.responseData === 'string') {
            return this.responseData;
        }
        if (this.dataAccess.getMetadata().getVersion() === '4.0') {
            if (this.statusCode === 204) {
                return;
            }
            if (this.isMinimalRepresentation) {
                this.statusCode = 204;
                this.addResponseHeader('preference-applied', 'return=minimal');
                return null;
            }
            else {
                this.addResponseHeader('content-type', 'application/json;odata.metadata=minimal;IEEE754Compatible=true');
                let outContext = this.context || '';
                if (outContext.indexOf('$metadata') === -1) {
                    outContext = `$metadata#${this.context}`;
                }
                let resultObject = {
                    '@odata.context': outContext
                };
                const metadataETags = this.dataAccess.getMetadata().getETag();
                if (metadataETags) {
                    resultObject['@odata.metadataEtag'] = metadataETags;
                }
                if (Array.isArray(this.responseData)) {
                    resultObject['@odata.count'] = this.dataCount;
                    resultObject.value = this.responseData;
                }
                else {
                    resultObject = { ...resultObject, ...this.responseData };
                }
                return JSON.stringify(resultObject);
            }
        }
        else {
            if (this.statusCode === 204) {
                return;
            }
            // V2
            const resultObject = { d: {} };
            if (Array.isArray(this.responseData)) {
                resultObject.d.__count = this.dataCount;
                resultObject.d.results = this.responseData;
            }
            else {
                resultObject.d = this.responseData;
            }
            return JSON.stringify(resultObject);
        }
    }
    addMessage(code, message, severity, target) {
        this.addCustomMessage({ code, message, numericSeverity: severity, target });
    }
    addCustomMessage(messageData) {
        this.messages.push(messageData);
    }
    addResponseHeader(headerName, headerValue) {
        this.responseHeaders[headerName] = headerValue;
    }
    setDataCount(dataCount) {
        this.dataCount = dataCount;
    }
}
exports.default = ODataRequest;
